src/plugins/grpc/server.js

```diff
@@ -0,0 +1,166 @@
+    'error.type': err.name
+  })
+}
+
+function createWrapHandler (grpc, tracer, config, handler) {
+  const configFields = getFilter(config, 'fields')
+  const configMetadata = getFilter(config, 'metadata')
+
+  return function wrapHandler (func) {
+    return function funcWithTrace (emitter, callback) {
+      const metadata = emitter.metadata
+      const request = emitter.request
+      const type = this.type
+      const isStream = type !== 'unary'
+      const childOf = tracer.extract(FORMAT_TEXT_MAP, metadata.getMap())
+      const span = tracer.startSpan('grpc.request', {
+        childOf,
+        tags: {
+          [Tags.SPAN_KIND]: Tags.SPAN_KIND_RPC_SERVER,
+          'grpc.method.name': handler,
+          'grpc.method.type': type,
+          'resource.name': handler.split('/').slice(-1),
+          'service.name': config.service || `${tracer._service}-grpc-server`
+        }
+      })
+
+      tracer.scopeManager().activate(span)
+
+      if (request) {
+        if (configMetadata && metadata) {
+          const values = configMetadata(metadata.getMap())
+
+          for (const key in values) {
+            span.setTag(`grpc.request.metadata.${key}`, values[key])
+          }
+        }
+
+        if (configFields) {
+          const values = configFields(request)
+
+          for (const key in values) {
+            span.setTag(`grpc.request.message.fields.${key}`, values[key])
+          }
+        }
+      }
+
+      // Finish the span if the call was cancelled.
+      emitter.on('cancelled', () => {
+        span.setTag('grpc.request.status', grpc.status.CANCELLED)
+        span.finish()
+      })
+
+      if (isStream) {
+        emitter.on('error', err => {
+          handleError(config, span, err)
+
+          span.finish()
+        })
+
+        // Finish the span of the response only if it was successful.
+        // Otherwise it'll be finished in the `error` listener.
+        emitter.on('finish', () => {
+          span.setTag('grpc.request.status', emitter.status.code)
+
+          if (emitter.status.code === 0) {
+            span.finish()
+          }
+        })
+
+        // Call the original stream request, without modification.
+        return func.apply(this, arguments)
+      }
+
+      // Call the unary request with a wrapped callback.
+      return func.call(this, emitter, (err, value, trailer, flags) => {
+        if (err) {
+          handleError(config, span, err)
+        }
+
+        if (trailer && configMetadata) {
+          const values = configMetadata(trailer.getMap())
+
+          for (const key in values) {
+            span.setTag(`grpc.response.metadata.${key}`, values[key])
+          }
+        }
+
+        // Unfortunately it doesn't seem possible to find the actual status code of the request at this point,
+        // so we're using these defaults until grpc adds proper support for server interceptors.
+        span.setTag('grpc.request.status', err ? grpc.status.UNKNOWN : grpc.status.OK)
```

Comment:
It's possible the error object contains a `code` property as well. It should be used when that's the case.

Reference: https://github.com/grpc/grpc-node/blob/633736225ac4f8ce05bd246ffe576616e9290615/packages/grpc-native-core/src/server.js#L53-L58

Reply:
Unfortunately `handleError` is only called later in the lifecycle. Essentially, it's called within the callback that I execute, [here](https://github.com/grpc/grpc-node/blob/633736225ac4f8ce05bd246ffe576616e9290615/packages/grpc-native-core/src/server.js#L595).

Reply:
Yes but if `handleError` sets `status.code` to the value of `error.code`, and `error.code` was already set at the point `handleError` is called it's possible that it's already set when this code runs as well.